### PR TITLE
Add additional fields to TxSettings for BasicStation

### LINKED ISSUE #358: Add additional fields to TxSettings of DownlinkMessage for BasicStation
**Summary:** 

Referring to [this comment](https://github.com/TheThingsNetwork/lorawan-stack/issues/120#issuecomment-476235240), we need to add a few additional parameters to the `TxSettings` that are necessary for BasicStation LNS protocol.

**Why do we need this?**

To support Downlinks in BasicStation gateways while using the GS Scheduler.

**What is already there? What do you see now?**

The required settings are not present.

**What is missing? What do you want to see?**

1. Add the following to `TxSettings` in `ttnpb`
-  Rx Window number indicating Rx1/Rx2
-  LoRaWAN Class
- AntennaIndex from the chosen path

2. Initialize them while processing downlinks.

**How do you propose to implement this?**

As explained under the above field.

**Environment:**

All

**What can you do yourself and what do you need help with?**

Can do this myself.

### PR DESCRIPTION
**Pull request by KrishnaIyer**
**Summary:**

Closes #358 

**Changes:**

- Add new fields to API.
- Use them in the GS downlink flow.

**Notes for Reviewers:**

Nil

**Release Notes: _(optional)_**

* Refactor `TxSettings` for better grouping of fields.

### CODE REVIEW

File: api/lorawan.proto [hunk 1]
```diff
@@ -272,6 +279,8 @@ message TxSettings {
   // Time of the gateway when the uplink message was received, or when the downlink message should be transmitted.
   // For downlink, this requires the gateway to have GPS time synchronization.
   google.protobuf.Timestamp time = 11 [(gogoproto.stdtime) = true];
+  // Additional information for the downlink from the original TxRequest; necessary for the BasicStation LNS gateways.
+  RequestInfo request_info = 12;
```
Comment (htdvisser) [inline api/lorawan.proto]: Why is this in a separate message? Can't we just add the fields to `TxSettings` directly, like we do with the other downlink-only `tx_power` field?

Comment (KrishnaIyer) [reply api/lorawan.proto]: Is there any advantage to adding them directly? Or is that a convention we follow?
I've used separate message only to group all the items necessary only for `BasicStation`.

Comment (htdvisser) [reply api/lorawan.proto]: Firstly, we don't tailor our API for specific gateways. Secondly, `antenna_index` may also be relevant for other multi-antenna gateways. 

The `rx_window` and `class` fields only seem relevant for gateways that have their own scheduler, so I don't quite understand why they need to be in the API. 

From the [comment leading to this PR](https://github.com/TheThingsNetwork/lorawan-stack/issues/120#issuecomment-476235240) I understand that we need to come up with an artificial `xtime`. Why do we need to know the RX window for that? Isn't it sufficient to have the Basic Station frontend set the chosen frequency/DR as RX1 (regardless), an `RxDelay` of 15 and `xtime` of `Timestamp - 15sec` (accounting for overflow)?

**Comment by KrishnaIyer** [comment]
Based on a better suggestion by @htdvisser , only `antenna_index` is added to `TxSettings`.

File: api/lorawan.proto [hunk 2]
```diff
@@ -272,6 +272,8 @@ message TxSettings {
   // Time of the gateway when the uplink message was received, or when the downlink message should be transmitted.
   // For downlink, this requires the gateway to have GPS time synchronization.
   google.protobuf.Timestamp time = 11 [(gogoproto.stdtime) = true];
+  // Index of the antenna on which the uplink was received and/or downlink must be sent.
+  uint32 antenna_index = 12;
```
Comment (johanstokking) [inline api/lorawan.proto]: `TxSettings` is specifically for transmission, from both the end device and gateway perspective. There cannot and should not be anything related to receiving here; that is `RxMetadata`.

It seems that we gradually added fields and set them along the way, while they are strictly not the transmission settings that were used by the transmitter.

I'm suggesting the following;

- Promote `gateway_channel_index` and `device_channel_index` to `UplinkMessage`
- Introduce a `TxSettings.Downlink` where we group fields that are specific for transmission from a gateway
   - `tx_power`
   - `antenna_index`
   - `invert_polarization`

Comment (KrishnaIyer) [reply api/lorawan.proto]: Ok done. I've rebased the commits to reflect only the newest proposal.

File: api/lorawan.proto [hunk 3]
```diff
@@ -240,6 +240,16 @@ message DataRate {
   };
 }
 
+// Fields that are specific for transmission on a gateway 
+message Downlink{
```
Comment (johanstokking) [inline api/lorawan.proto]: Don't make this top-level but declare in `TxSettings`, as they are local to that

File: pkg/ttnpb/lorawan_populate.go [hunk 4]
```diff
@@ -317,7 +317,13 @@ func NewPopulatedTxRequest(r randyLorawan, easy bool) *TxRequest {
 }
 
 func NewPopulatedTxSettings(r randyLorawan, easy bool) *TxSettings {
-	out := &TxSettings{}
+
```
Comment (johanstokking) [inline pkg/ttnpb/lorawan_populate.go]: Remove newline

File: pkg/ttnpb/udp/translation_test.go [hunk 5]
```diff
@@ -302,16 +305,20 @@ func TestDownlinkRoundtrip(t *testing.T) {
...
+				},
+				Timestamp: 188700000,
 			},
 		},
 		RawPayload: []byte{0x7d, 0xf3, 0x8e},
 	}
 	tx, err := udp.FromDownlinkMessage(expected)
 	a.So(err, should.BeNil)
 
+	fmt.Println(tx.Powe)
```
Comment (johanstokking) [inline pkg/ttnpb/udp/translation_test.go]: Remove

File: api/lorawan.proto [hunk 6]
```diff
@@ -254,24 +254,26 @@ message TxSettings {
...
+  bool enable_crc = 5 [(gogoproto.customname) = "EnableCRC"];
   // Timestamp of the gateway concentrator when the uplink message was received, or when the downlink message should be transmitted (microseconds).
   // On downlink, set timestamp to 0 and time to null to use immediate scheduling.
-  uint32 timestamp = 10;
+  uint32 timestamp = 6;
   // Time of the gateway when the uplink message was received, or when the downlink message should be transmitted.
   // For downlink, this requires the gateway to have GPS time synchronization.
-  google.protobuf.Timestamp time = 11 [(gogoproto.stdtime) = true];
+  google.protobuf.Timestamp time = 7 [(gogoproto.stdtime) = true];
+  // Fields that are specific for transmission on a gateway
```
Comment (johanstokking) [inline api/lorawan.proto]: `Transmission settings for downlink.`

File: api/lorawan.proto [hunk 7]
```diff
@@ -254,24 +254,26 @@ message TxSettings {
...
-  uint32 timestamp = 10;
+  uint32 timestamp = 6;
   // Time of the gateway when the uplink message was received, or when the downlink message should be transmitted.
   // For downlink, this requires the gateway to have GPS time synchronization.
-  google.protobuf.Timestamp time = 11 [(gogoproto.stdtime) = true];
+  google.protobuf.Timestamp time = 7 [(gogoproto.stdtime) = true];
+  // Fields that are specific for transmission on a gateway 
+  Downlink downlink = 8;
+
+  // Fields that are specific for transmission on a gateway
```
Comment (johanstokking) [inline api/lorawan.proto]: - `Transmission settings for downlink.`
- Declare before use of field, like other embedded messages

Comment (KrishnaIyer) [reply api/lorawan.proto]: Actually I used the `message MACCommand` as reference on `L366`

**Comment by KrishnaIyer** [comment]
done. @htdvisser please take a look.

**Pull request merged**